File: apps/host/src/utils/xbitAvatar/ColorWheel.ts
import {cycleIndex} from "@/utils/array.ts"
import {hsl2hex} from "@/utils/xbitAvatar/ColorUtil.ts";

/**
 * ColorWheel is color theory in art painting, please Google to learn & understand color wheel first
 */
export class ColorWheel {
  configs: ColorWheelConfig = {
    hues: [
      // 0, // Red
      // 30, // Red-Orange
      // 60, // Yellow
      // 90, // Yellow-Green
      // 120, // Green
      // 150, // Green-Cyan
      // 180, // Cyan
      // 210, // Cyan-Blue
      // 240, // Blue
      // 270, // Blue-Magenta
      // 300, // Magenta
      // 330 // Magenta-Red
    ],
    lightnessLevels: [],
    saturate: 1,
  }

  // output
  colors: Color[] = [];

  /**
   * @param hueLevelsCount auto generate hue steps base on hue level count, eg: count=6 will generate 6 colors in the color wheel
   * @param saturate make more color saturate, range: 0-1
   * @param lightnessLevels array of range: 0-1, it's how many ring we made in the color wheel, must be in decreasing order
   */
  constructor(
    hueLevelsCount: number,
    saturate: number,
    lightnessLevels: number[],
  ) {
    const hues: number[] = [];
    for (let i = 0; i < hueLevelsCount; i++) {
      hues.push(Math.floor(i * 360 / hueLevelsCount));
    }

    this.configs.hues = hues;
    this.configs.lightnessLevels = lightnessLevels;
    this.configs.saturate = saturate;
    this.colors = ColorWheel.generateColorWheelColors(hues, saturate, lightnessLevels);
  }

  /**
   * Generate color pallet base on color wheel
   * amount returned = 12 hue * count sat levels
   * @returns Array of color objects with hsl, hex and data properties
   */
  static generateColorWheelColors(
    hues: number[] = [],
    saturate: number = 0.75,
    lightnessLevels: number[] = [0.75, 0.5, 0.25],
  ): Color[] {
    const colors: Color[] = [];

    for (let i = 0, c = hues.length; i < c; i++) {
      const hue = hues[i]
      for (const lightness of lightnessLevels) {
        // The saturation is kept at a high value for vibrant colors
        const color: Color = {
          // hsl: `hsl(${hue}, ${saturate * 100}%, ${lightness * 100}%)`,
          hex: hsl2hex(hue, saturate, lightness),
          data: {
            h: hue, s: saturate, l: lightness,
            hi: i,
            i: colors.length,
          },
        }
        colors.push(color)
      }
    }

    return colors
  }

  getColors(): Color[] {
    return this.colors;
  }

  getHexColors(): string[] {
    return this.colors.map(i => i.hex);
  }

  // brighter(color) {
  //
  // }
  // darker(color) {
  //
  // }

  // return Color type, this ensure returned color have enough contrast compared to srcColor
  getContrastColorOf(srcColorIdx: number, srcColorWheel: ColorWheel, seed: number, _debugData?: any): Color {
    // hue is the same config between all color wheel, so hueIdx is the same for src & target
    const hueIdx = srcColorWheel.getHueIdx(srcColorIdx)
    // same hue or hue + 1 is not enough contrast, so need to hue += 2
    const shiftDegree = 180
    const hueIdxDelta = Math.floor(this.configs.hues.length * shiftDegree / 360) // must shift hue +60 degree, means 60/360 = 1/6 color wheel
    const lc = this.configs.lightnessLevels.length;
    const palletColorDelta = hueIdx + hueIdxDelta * lc + seed % lc;
    const targetColorIdx = Math.floor(srcColorIdx / srcColorWheel.configs.hues.length) + (srcColorIdx % srcColorWheel.configs.hues.length)
    const contrastColorIdx = cycleIndex(targetColorIdx, palletColorDelta, this.colors)
    const contrastColor = this.colors[contrastColorIdx]

    // Turn verifyHslContrast on for debug only
    // const debug = 1
    // if (debug) {
    //   const srcColor = srcColorWheel.colors[srcColorIdx]
    //   const verify = verifyHslContrast(srcColor.data, contrastColor.data)
    //   if (!verify.verified.xbit_logo_bg) {
    //     console.log(`not enough contrast: `, {
    //       verifyHslContrast: verify,
    //       srcColorIdx,
    //       contrastColorIdx,
    //       srcColor,
    //       contrastColor,
    //       seed,
    //       hueIdxDelta,
    //       palletColorDelta,
    //       _debugData,
    //     })
    //   }
    // }

    return contrastColor
  }

  getHueIdx(colorIdx: number): number {
    return this.colors[colorIdx].data.hi
  }
}

export interface ColorWheelConfig {
  hues: number[]; // in degree, 0-360 degree in the color wheel
  lightnessLevels: number[]; // range: 0-1
  saturate: number; // range 0-1
}

export interface ColorData {
  h: number;
  s: number;
  l: number;
  hi: number; // aka hue index
  i: number; // index of color in the pallet arr
}

export interface Color {
  hex: HexColor;
  data: ColorData;
}

export type HexColor = string

File: apps/host/src/utils/xbitAvatar/FlutterDart/avatar/ColorWheel.dart
import 'dart:math';
import 'ColorUtil.dart';

typedef HexColor = String;

/// ColorWheel is color theory in art painting, please Google to learn & understand color wheel first
class ColorWheel {
  ColorWheelConfig configs = ColorWheelConfig(
    hues: [],
    lightnessLevels: [],
    saturate: 1.0,
  );

  // output
  List<Color> colors = [];

  /// @param hueLevelsCount auto generate hue steps base on hue level count, eg: count=6 will generate 6 colors in the color wheel
  /// @param saturate make more color saturate, range: 0-1
  /// @param lightnessLevels array of range: 0-1, it's how many ring we made in the color wheel, must be in decreasing order
  ColorWheel(
      int hueLevelsCount,
      double saturate,
      List<double> lightnessLevels,
      ) {
    final List<double> hues = [];
    for (int i = 0; i < hueLevelsCount; i++) {
      hues.add((i * 360 / hueLevelsCount).floorToDouble());
    }

    configs.hues = hues;
    configs.lightnessLevels = lightnessLevels;
    configs.saturate = saturate;
    colors = ColorWheel.generateColorWheelColors(hues, saturate, lightnessLevels);
  }

  /// Generate color pallet base on color wheel
  /// amount returned = 12 hue * count sat levels
  /// @returns Array of color objects with hsl, hex and data properties
  static List<Color> generateColorWheelColors(
      List<double> hues,
      double saturate,
      List<double> lightnessLevels,
      ) {
    final List<Color> colors = [];

    for (int i = 0; i < hues.length; i++) {
      final hue = hues[i];
      for (final lightness in lightnessLevels) {
        final color = Color(
          hex: ColorUtil.hsl2hex(hue, saturate, lightness),
          data: ColorData(
            h: hue,
            s: saturate,
            l: lightness,
            hi: i,
            i: colors.length,
          ),
        );
        colors.add(color);
      }
    }

    return colors;
  }

  List<Color> getColors() {
    return colors;
  }

  List<String> getHexColors() {
    return colors.map((i) => i.hex).toList();
  }

  /// return Color type, this ensure returned color have enough contrast compared to srcColor
  Color getContrastColorOf(
      int srcColorIdx,
      ColorWheel srcColorWheel,
      int seed, [
        dynamic debugData,
      ]) {
    // hue is the same config between all color wheel, so hueIdx is the same for src & target
    final hueIdx = srcColorWheel.getHueIdx(srcColorIdx);
    // same hue or hue + 1 is not enough contrast, so need to hue += 2
    const shiftDegree = 180;
    final hueIdxDelta = (configs.hues.length * shiftDegree / 360).floor();
    final palletColorDelta = hueIdx +
        hueIdxDelta * configs.lightnessLevels.length +
        seed % configs.lightnessLevels.length;
    final targetColorIdx = (srcColorIdx / srcColorWheel.configs.hues.length).floor() +
        (srcColorIdx % srcColorWheel.configs.hues.length);
    final contrastColorIdx = _cycleIndex(targetColorIdx, palletColorDelta, colors);
    final contrastColor = colors[contrastColorIdx];

    return contrastColor;
  }

  int getHueIdx(int colorIdx) {
    return colors[colorIdx].data.hi;
  }

  /// Helper function to cycle through array indices
  int _cycleIndex(int current, int delta, List<Color> array) {
    return (current + delta) % array.length;
  }
}

class ColorWheelConfig {
  List<double> hues; // in degree, 0-360 degree in the color wheel
  List<double> lightnessLevels; // range: 0-1
  double saturate; // range 0-1

  ColorWheelConfig({
    required this.hues,
    required this.lightnessLevels,
    required this.saturate,
  });
}

class ColorData {
  double h;
  double s;
  double l;
  int hi; // aka hue index
  int i; // index of color in the pallet arr

  ColorData({
    required this.h,
    required this.s,
    required this.l,
    required this.hi,
    required this.i,
  });
}

class Color {
  HexColor hex;
  ColorData data;

  Color({
    required this.hex,
    required this.data,
  });
}
